### PR TITLE
Add AWS collector and recipe

### PR DESCRIPTION
**Pull request by giovannt0**
Add a new `aws_forensics` recipe and a new `AWSCollector` collector, that offers similar functionalities to the `gcp_forensics` recipe and its corresponding `GoogleCloudCollector` collector.

Marking as draft until libcloudforensics new pip release is out.

Signed-off-by: Theo Giovanna <gtheo@google.com>

### CODE REVIEW

File: dftimewolf/lib/collectors/aws.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+      cpu_cores (int): Optional. The number of CPU cores to use for the
+          analysis VM. Default is 16.
+      ami (str): Optional. The Amazon Machine Image ID to use to create the
+          analysis VM. Default is Ubuntu 18.04 TLS.
+    """
+
+    if not (remote_profile_name and remote_zone):
+      self.state.AddError('You must specify "remote_profile_name" and "zone" '
+                          'parameters', critical=True)
+
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:134]: You need to return right after adding the error (pending https://github.com/log2timeline/dftimewolf/issues/80)

File: dftimewolf/lib/collectors/aws.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    return volumes_to_copy
+
+  def _FindNextAvailableDeviceName(self):
+    """Determine the next available device name to attach volumes to the VM.
+
+    AWS recommends using device names that are within /dev/sd[f-p][1-6].
+
+    Returns:
+      str: A device name.
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:239]: Or an empty string if the device name could not be obtained?

File: dftimewolf/lib/collectors/aws.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    self.incident_id = incident_id
+    self.remote_instance_id = remote_instance_id
+
+    self.volume_ids = volume_ids.split(',') if volume_ids else []
+    self.all_volumes = all_volumes
+    self.analysis_zone = analysis_zone or remote_zone
+    self.analysis_profile_name = analysis_profile_name or remote_profile_name
+
+    # See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html
+    self.device_names = [chr(i) for i in range(ord('f'), ord('p') + 1)]
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:149]: You can also do

```python
self.device_names = list('fghijklmnop')
```

I'd also rename this attribute to `device_suffixes`

File: dftimewolf/lib/collectors/aws.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    """
+    volumes = []
+    for volume_id in volume_ids:
+      try:
+        volumes.append(self.source_account.GetVolumeById(volume_id))
+      except RuntimeError:
+        self.state.AddError(
+            'Volume "{0:s}" was not found in AWS account {1:s}'.format(
+                volume_id, self.remote_profile_name),
+            critical=True)
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:184]: If the error is critical, then you will want to return an empty list, and check for the return value of this in the caller (I know, I should really fix #80)

File: dftimewolf/lib/collectors/aws.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    return volumes
+
+  def _GetVolumesFromInstance(self, instance_id, all_volumes):
+    """Gets volumes to copy based on an instance name.
+
+    Args:
+      instance_id (str): Instance ID of the instance to get the volumes from.
+      all_volumes (bool): If set, get all volumes attached to the instance. If
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:192]:
```suggestion
      all_volumes (bool): If True, get all volumes attached to the instance. If
```

File: dftimewolf/lib/collectors/aws.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    except RuntimeError as exception:
+      self.state.AddError(str(exception), critical=True)
+      return []
+
+    if all_volumes:
+      return list(remote_instance.ListVolumes().values())
+    return [remote_instance.GetBootVolume()]
+
+  def _FindVolumesToCopy(self):
+    """Determines which volumes to copy depending on object attributes.
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:209]:
```suggestion
    """Determines which volumes to copy depending on the collector's attributes.
```

**Comment by giovannt0** [comment]
Placing this on hold until https://github.com/google/cloud-forensics-utils/issues/159 is resolved, since it will likely change the collector.

**Comment by giovannt0** [comment]
@tomchop PTAL, I updated the code to work with the new libcloudforensics release and addressed your comments. Note that the pinned dependency for libcloudforensics should be updated first (to 20200630 or more recent).

File: data/recipes/aws_forensics.json [hunk 7]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "aws_forensics",
+  "short_description": "Copies volume from an AWS account to an analysis VM.",
```
Comment (tomchop) [inline data/recipes/aws_forensics.json:3]:
```suggestion
  "short_description": "Copies a volume from an AWS account to an analysis VM.",
```

File: dftimewolf/lib/collectors/aws.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+  Attributes:
+    remote_profile_name (str): The AWS account in which the volume(s) exist(s).
+        This is the profile name that is defined in your AWS credentials file.
+    remote_zone (str): The AWS zone in which the source volume(s) exist(s).
+    source_account (AWSAccount): The AWSAccount object that represents the
+        source account in which the volume(s) exist(s).
+    incident_id (str): Incident identifier used to name the analysis VM.
+    remote_instance_id (str): Instance ID that needs forensicating.
+    volume_ids (list[str]): List of volume ids to copy.
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:23]:
```suggestion
    volume_ids (list[str]): List of volume IDs to copy.
```

File: dftimewolf/lib/collectors/aws.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+          your AWS credentials file.
+      analysis_zone (str): Optional. The AWS zone in which to create the VM.
+          If not specified, the VM will be created in the same zone where the
+          volume(s) exist(s).
+      boot_volume_size (int): Optional. The size (in GB) of the boot volume
+          for the analysis VM. Default is 50 GB.
+      cpu_cores (int): Optional. The number of CPU cores to use for the
+          analysis VM. Default is 16.
+      ami (str): Optional. The Amazon Machine Image ID to use to create the
+          analysis VM. Default is Ubuntu 18.04 TLS.
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:128]: It seems like the default parameter here is None. So maybe rephrase this to say `if not specified, will default to selecting Ubuntu 18.04 LTS`.

File: dftimewolf/lib/collectors/aws.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+          self.remote_zone,
+          dst_zone=self.analysis_zone,
+          volume_id=volume.volume_id,
+          src_profile=self.remote_profile_name,
+          dst_profile=self.analysis_profile_name)
+      self.analysis_vm.AttachVolume(
+          new_volume, self._FindNextAvailableDeviceName())
+      print('Volume {0:s} successfully copied to {1:s}'.format(
+          volume.volume_id, new_volume.volume_id))
+      self.state.output.append((self.analysis_vm.name, new_volume))
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:74]: Since #279 we're not using input / output anymore. Please replace this with a a `ForensicsVM` attribute container like in [gcloud.py](https://github.com/log2timeline/dftimewolf/blob/master/dftimewolf/lib/collectors/gcloud.py#L70)

File: dftimewolf/lib/collectors/aws.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    self.incident_id = incident_id
+    self.remote_instance_id = remote_instance_id
+
+    self.volume_ids = volume_ids.split(',') if volume_ids else []
+    self.all_volumes = all_volumes
+    self.analysis_zone = analysis_zone or remote_zone
+    self.analysis_profile_name = analysis_profile_name or remote_profile_name
+
+    # See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html
+    self.device_suffixes = list('fghijklmnop')
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:150]: Shouldn't this go in `__init__`?

File: dftimewolf/lib/collectors/aws.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+    """Determines which volumes to copy depending on the collector's attributes.
+
+    Returns:
+      list[AWSVolume]: A list of the volumes to copy.
+    """
+    if not (self.remote_instance_id or self.volume_ids):
+      self.state.AddError(
+          'You need to specify at least an instance name or volume ids to copy',
+          critical=True)
+      return []
```
Comment (tomchop) [inline dftimewolf/lib/collectors/aws.py:220]: This should go in `SetUp`. We want to be able to fail early if these parameters are not provided. (Should also be changed in `gcloud.py`, #297 )

Comment (giovannt0) [reply dftimewolf/lib/collectors/aws.py]: Thanks, opened #298 for `gcloud.py`
